apps/dcos_dns/src/dcos_dns_handler.erl

```diff
@@ -177,90 +177,106 @@ resolve_loop(Begin, Workers, Fun, FailMsg, Zone) ->
                 ok ->
-                    udp_worker(StartTime, Pid, Socket, Upstream);
-                {error, Error} ->
-                    lager:warning("DNS worker ~p failed with ~p", [Upstream, Error])
+                    udp_worker_receive(Pid, {IP, Port}, Socket);
+                {error, Reason} ->
+                    {error, Reason}
             end;
-        {error, Error} ->
-            lager:warning("DNS worker ~p failed with ~p", [Upstream, Error])
+        {error, Reason} ->
+            {error, Reason}
     end.
 
--spec(udp_worker(integer(), pid(), gen_udp:socket(), upstream()) -> ok).
-udp_worker(StartTime, Pid, Socket, Upstream = {IP, Port}) ->
+-spec(udp_worker_receive(pid(), upstream(), gen_udp:socket()) ->
+    {ok, binary()} | {error, atom()}).
+udp_worker_receive(Pid, {IP, Port}, Socket) ->
     MonRef = erlang:monitor(process, Pid),
     try
         receive
-            {'DOWN', MonRef, process, _Pid, normal} ->
-                ok;
-            {'DOWN', MonRef, process, _Pid, Reason} ->
-                exit(Reason);
+            {'DOWN', MonRef, process, _Pid, _Reason} ->
+                {error, down};
             {udp, Socket, IP, Port, Response} ->
-                Pid ! {reply, self(), Response},
-                report_latency([?MODULE, Upstream, latency], StartTime);
+                {ok, Response};
             {timeout, Pid} ->
-                lager:warning("DNS worker ~p timed out", [Upstream])
+                {error, timeout}
         after ?TIMEOUT ->
-            ok
+            {error, timeout}
         end
     after
         gen_udp:close(Socket)
     end.
 
--spec(tcp_worker(pid(), upstream(), binary()) -> ok).
-tcp_worker(Pid, Upstream = {IP, Port}, Request) ->
+-spec(tcp_worker(pid(), upstream(), binary()) ->
+   {ok, binary()} | {error, atom()}).
+tcp_worker(Pid, {IP, Port}, Request) ->
     StartTime = erlang:monotonic_time(millisecond),
     Opts = [{active, once}, binary, {packet, 2}, {send_timeout, ?TIMEOUT}],
     case gen_tcp:connect(IP, Port, Opts, ?TIMEOUT) of
         {ok, Socket} ->
             case gen_tcp:send(Socket, Request) of
                 ok ->
-                    tcp_worker(StartTime, Pid, Socket, Upstream);
-                {error, Error} ->
-                    lager:warning("DNS worker [tcp] ~p failed with ~p", [Upstream, Error])
+                    tcp_worker_receive(StartTime, Pid, Socket);
+                {error, Reason} ->
+                    {error, Reason}
             end;
-        {error, Error} ->
-            lager:warning("DNS worker [tcp] ~p failed with ~p", [Upstream, Error])
+        {error, Reason} ->
+            {error, Reason}
     end.
 
--spec(tcp_worker(integer(), pid(), gen_tcp:socket(), upstream()) -> ok).
-tcp_worker(StartTime, Pid, Socket, Upstream) ->
+-spec(tcp_worker_receive(integer(), pid(), gen_tcp:socket()) ->
+   {ok, binary()} | {error, atom()}).
+tcp_worker_receive(StartTime, Pid, Socket) ->
```

Comment:
It looks like `udp_worker_receive/3` and `tcp_worker_receive/3` have almost the same logic, let merge them into one function?